tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -296,9 +329,23 @@ StorageManager::array_open_for_writes(Array* array) {
       load_array_schemas(array->array_directory(), *array->encryption_key());
   RETURN_NOT_OK_TUPLE(st_schemas, nullopt, nullopt);
 
-  // This library should not be able to write to newer-versioned arrays
+  // If building experimentally, this library should not be able to
+  // write to newer-versioned or older-versioned arrays
+  // Else, this library should not be able to write to newer-versioned arrays
   // (but it is ok to write to older arrays)
   auto version = array_schema_latest.value()->version();
+  if constexpr (is_experimental_build) {
+    if (version != constants::format_version) {
+      std::stringstream err;
+      err << "Cannot open array for writes; Array format version (";
+      err << version;
+      err << ") is not the library format version (";
+      err << constants::format_version << ")";
+      return {logger_->status(Status_StorageManagerError(err.str())),
+              nullopt,
+              nullopt};
+    }
+  }
```

Comment:
```suggestion
  } else {
    if (version > constants::format_version) {
    ...
    }
  }
```

Reply:
Not sure I follow this one - let's discuss

Reply:
(the goal is to _disallow_ writing to any older format version when `is_experimental_build` is true, because we don't want to create arrays which are unreadable from a release-build)